pallet_stream: double map to provide latest state for a given 'identifer:hash'

amarts opened the pull request:
This is a sample on adding a doublemap to get the status of given hash / digest.

amarts commented:
wanted to dust-off my coding on cord. @smohan-dw I thought having just the status is good enough in this query, as identifier's query would be having all the other details. This will make verification proper IMO.

Yet to test, but meantime wanted feedback if I am very much off the track.

Updates #239

smohan-dw commented on pallets/stream/src/lib.rs:
```diff
@@ -267,12 +281,14 @@ pub mod pallet {
 				},
 			);
 
+			<UniqueStreams<T>>::insert(&identifier, stream_digest, false);
```
+1. this can become the revocation registry as well. We can think of removing the revoke field from the StreamEntry and make this structure the state proof.

amarts replied:
True. Just that we have to make it clear, (or do 2 queries from SDK) that one is state check, and another is identifier latest details check.

amarts replied:
moved 'creator' and 'revoked_by' also into the structure.

roynrishingha reviewed (commented):
@amarts is not this PR same as my PR?

amarts commented:
Similar, not same. 

Changes:

* Structure of what is stored.
* No extra fetch for digest, and reuse the existing.
* Handle all cases for keeping the <identity:hash> key's status up-to-date.

amarts committed:
pallet_stream: fix the storage to manage digest based query

by adding a StorageDoubleMap we can now query specific digest of the identifier, and its state.

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts commented:
Few updates:

* Name the storage as 'Attestations'.
* Add 'creator' and 'revoked_by' fields into attestation, as they can be different accounts which can be performing operations on same identifiers.
* Removed 'revoked' flag from Identifier (which means, the query should change in cord.js verification methods.
* stream::remove(), removes all entries in the given identifier.

smohan-dw commented on pallets/stream/src/lib.rs:
```diff
@@ -135,6 +140,20 @@ pub mod pallet {
+		StreamIdOf,
+		Blake2_128Concat,
+		StreamDigestOf<T>,
+		StreamAttestationEntryOf<T>,
```
This can be StreamEnty (or rename it to AttestationDetails). With that, We can change Streams to StreamIdentifier, which maps an identifier to the latest digest.

amarts replied:
updated this. 

Other storage update IMO can come in as separate PR. So, that review will be cleaner.

amarts committed:
review comments: AttestationDetails

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
clippy fixes: remove revoke from Streams

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
few more changes as per the review

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
fix cargo test

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
fixes in benchmarking

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
clippy local run passing

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
fix benchmark tests

Signed-off-by: Amar Tumballi <amar@dhiway.com>

amarts committed:
clippy error fix

Signed-off-by: Amar Tumballi <amar@dhiway.com>

smohan-dw commented:
LGTM. Please hold the merge for a few hours. Almost completing the SDK update

smohan-dw reviewed (commented):
LGTM. We should get the SDK updated before the merge. SDK changes are required as the storage parameters are changed.

amarts commented:
> LGTM. Please hold the merge for a few hours. Almost completing the SDK update

@smohan-dw done - https://github.com/dhiway/cord.js/pull/132
